File: Niv1984/app/src/main/java/com/aaronicsubstances/niv1984/etc/BookTextViewUtils.java
package com.aaronicsubstances.niv1984.etc;

import android.app.Activity;
import android.content.Context;
import android.net.Uri;
import android.webkit.ConsoleMessage;
import android.webkit.MimeTypeMap;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.aaronicsubstances.niv1984.R;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.regex.Pattern;

/**
 * Created by Aaron on 12/2/2017.
 */

public class BookTextViewUtils {
    public static final String LAUNCH_URL = "http://localhost";
    public static final int DEFAULT_ZOOM_INDEX = 1;
    public static final int DEFAULT_LINE_HEIGHT_INDEX = 2;
    private static final Pattern HTML_SUFFIX_PATTERN = Pattern.compile("\\.html\\d*$");

    private static final Logger LOGGER = LoggerFactory.getLogger(BookTextViewUtils.class);

    public static void configureBrowser(final Activity context, WebView browser,
                                        CustomWebPageEventListener listener) {
        WebSettings webSettings = browser.getSettings();
        webSettings.setJavaScriptEnabled(true);
        webSettings.setSaveFormData(false);
        webSettings.setCacheMode(WebSettings.LOAD_NO_CACHE);
        webSettings.setLayoutAlgorithm(WebSettings.LayoutAlgorithm.TEXT_AUTOSIZING);

        final SharedPrefsManager sharedPrefsManager = new SharedPrefsManager(context);

        browser.setWebChromeClient(new WebChromeClient() {

            @Override
            public boolean onConsoleMessage(ConsoleMessage consoleMessage) {
                LOGGER.error("{}: {} (at {}:{})",
                        consoleMessage.messageLevel(), consoleMessage.message(),
                        consoleMessage.sourceId(), consoleMessage.lineNumber());
                return true;
            }
        });

        browser.setWebViewClient(new WebViewClient(){

            @Override
            public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
                if (request.getUrl().getHost().equals("localhost")) {
                    try {
                        return serve(context, request.getUrl(), sharedPrefsManager);
                    }
                    catch (Exception ex) {
                        LOGGER.error("Could not serve {}: {}", request.getUrl(), ex);
                    }
                }
                return super.shouldInterceptRequest(view, request);
            }

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                if (url.startsWith(LAUNCH_URL)) {
                    return false;
                }
                return true;
            }
        });

        browser.addJavascriptInterface(new BibleJs(context, listener),
                BibleJs.NAME);
    }

    /**
     * Serves static assets at localhost urls.
     *
     * @param context
     * @param uri localhost url
     * @param sharedPrefsManager
     * @return WebResourceResponse
     * @throws IOException if an error occurs (e.g. url doesn't point to any asset file).
     */
    public static WebResourceResponse serve(Context context, Uri uri,
                                            SharedPrefsManager sharedPrefsManager) throws IOException {
        String assetPath = uri.getPath();
        if (assetPath.startsWith("/")) {
            assetPath = assetPath.substring(1);
        }
        if (assetPath.isEmpty() || assetPath.endsWith("/")) {
            assetPath += "index.html";
        }

        InputStream assetStream;
        if (assetPath.equals("favicon.ico")) {
            LOGGER.debug("favicon.ico request found. Returning empty response.");
            assetStream = new ByteArrayInputStream(new byte[0]);
        }
        else {
            String prevAssetPath = assetPath;
            assetPath = HTML_SUFFIX_PATTERN.matcher(prevAssetPath).replaceFirst(".html");
            if (prevAssetPath.equals(assetPath))
            {
                LOGGER.debug("Fetching asset '{}' ...", assetPath);
            }
            else
            {
                LOGGER.debug("Fetching asset '{}' for path '{}' ...", assetPath, prevAssetPath);
            }
            assetStream = context.getAssets().open(assetPath);
        }

        // Before serving do something special for css/base.css
        // by appending zoom and line height information.
        if (assetPath.equals("css/base.css")) {
            int zoomLevelIndex = sharedPrefsManager.getZoomLevelIndex();
            if (zoomLevelIndex < 0) {
                zoomLevelIndex = DEFAULT_ZOOM_INDEX;
            }
            int lineHeightIndex = sharedPrefsManager.getLineHeightIndex();
            if (lineHeightIndex < 0) {
                lineHeightIndex = DEFAULT_LINE_HEIGHT_INDEX;
            }
            if (zoomLevelIndex != DEFAULT_ZOOM_INDEX || lineHeightIndex != DEFAULT_LINE_HEIGHT_INDEX) {
                LOGGER.debug("Appending custom css...");
                String originalCss = Utils.toString(assetStream);
                assetStream.close();
                String zoom = context.getResources().getStringArray(R.array.zoom_entries_slim)[zoomLevelIndex];
                // for some reason percentages were not working with line-height of 250% upwards,
                // so had to use floating-points.
                double lineHeight = 1.25 + 0.25 * lineHeightIndex;
                assetStream = new ByteArrayInputStream(String.format("%s%nbody { font-size: %s; line-height: %s; }",
                        originalCss, zoom, lineHeight).getBytes());
            }
        }

        String mimeType = getMimeType(assetPath);
        LOGGER.debug("Returning {} response from {}...", mimeType, assetPath);
        WebResourceResponse staticResponse = new WebResourceResponse(mimeType,
                "utf-8", assetStream);
        return staticResponse;
    }

    public static String resolveUrl(String rootRelativeUrl, String... queryParams) {
        StringBuilder launchUrl = new StringBuilder(LAUNCH_URL);
        if (launchUrl.charAt(launchUrl.length() - 1) != '/') {
            launchUrl.append("/");
        }
        if (rootRelativeUrl.startsWith("/")) {
            rootRelativeUrl = rootRelativeUrl.substring(1);
        }
        launchUrl.append(rootRelativeUrl);
        if (queryParams != null) {
            for (int i = 0; i < queryParams.length; i += 2) {
                String q = queryParams[i];
                String v = null;
                if (i+1 < queryParams.length) {
                    v = queryParams[i + 1];
                }
                if (i > 0) {
                    launchUrl.append("&");
                } else {
                    launchUrl.append("?");
                }
                launchUrl.append(Uri.encode(q));
                if (v != null) {
                    launchUrl.append("=");
                    launchUrl.append(Uri.encode(v));
                }
            }
        }
        return launchUrl.toString();
    }

    public static String getMimeType(String path) {
        String fileExtension = getFileExtension(path).toLowerCase();
        switch (fileExtension) {
            case "html":
                return "text/html";
            case "css":
                return "text/css";
            case "js":
                return "text/javascript";
            case "png":
                return "image/png";
            case "svg":
                return "image/svg+xml";
            case "woff":
                return "application/font-woff";
            case "eot":
                return "application/vnd.ms-fontobject";
            case "ttf":
                return "application/font-sfnt";
            case "otf":
                return "application/font-sfnt";
            case "ico":
                return "image/vnd.microsoft.icon";
            default:
                return MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension);
        }
    }

    private static String getFileExtension(String path) {
        int lastPeriodIndex = path.lastIndexOf('.');
        if (lastPeriodIndex != -1) {
            return path.substring(lastPeriodIndex+1);
        }
        return "";
    }
}

File: Niv1984-Plus/app/src/main/java/com/aaronicsubstances/niv1984/ui/about/BookTextViewUtils.kt
package com.aaronicsubstances.niv1984.ui.about

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Build
import android.util.Base64
import android.view.View
import android.webkit.*
import android.widget.ProgressBar
import androidx.core.content.ContextCompat.startActivity
import com.aaronicsubstances.niv1984.data.FirebaseFacade
import kotlinx.coroutines.runBlocking
import org.slf4j.LoggerFactory
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream


object BookTextViewUtils {
    const val LAUNCH_URL = "http://localhost"
    const val ASSETS_PREFIX = "assets/"
    private val LOGGER = LoggerFactory.getLogger(BookTextViewUtils::class.java)

    fun configureBrowser(context: Activity, browser: WebView, progressBar: ProgressBar?) {
        val webSettings = browser.settings
        webSettings.javaScriptEnabled = true
        webSettings.javaScriptCanOpenWindowsAutomatically = true
        webSettings.cacheMode = WebSettings.LOAD_NO_CACHE
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            webSettings.layoutAlgorithm = WebSettings.LayoutAlgorithm.TEXT_AUTOSIZING
        }
        browser.webChromeClient = object : WebChromeClient() {
            override fun onConsoleMessage(consoleMessage: ConsoleMessage): Boolean {
                LOGGER.error(
                    "{}: {} (at {}:{})",
                    consoleMessage.messageLevel(), consoleMessage.message(),
                    consoleMessage.sourceId(), consoleMessage.lineNumber()
                )
                return true
            }

            override fun onProgressChanged(view: WebView?, progress: Int) {
                progressBar?.progress = progress * 100
            }
        }
        browser.webViewClient = object : WebViewClient() {

            /* This function is strictly not needed in our case, but is
               kept here to document our findings about it.
               It is called when links are clicked. Returning true prevents
               further processing. By default it returns false, which means it
               will go on to call shouldInterceptRequest.
             */
            override fun shouldOverrideUrlLoading(
                view: WebView,
                url: String
            ): Boolean {
                if (url.contains("https://mywebsite.domain.com")) {
                    val i = Intent(Intent.ACTION_VIEW, Uri.parse(url))
                    context.startActivity(i)
                    return true
                } else {
                    return false
                }
            }

            override fun shouldInterceptRequest(
                view: WebView,
                url: String
            ): WebResourceResponse? {
                if (url.startsWith(LAUNCH_URL)) {
                    try {
                        return serve(context, url)
                    } catch (t: Exception) {
                        LOGGER.error("Could not serve $url: ", t)
                    }
                }
                // by default null is returned just like the default implementation,
                // to indicate that the usual process of
                // making an http request should be followed through.
                return null
            }

            override fun onPageStarted(
                view: WebView?,
                url: String?,
                favicon: Bitmap?
            ) {
                super.onPageStarted(view, url, favicon)
                progressBar?.visibility = View.VISIBLE
                progressBar?.progress = 0
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                super.onPageFinished(view, url)
                progressBar?.visibility = View.GONE
            }
        }
    }

    /**
     * Serves static assets at localhost urls.
     *
     * @param context
     * @param url localhost url
     * @return WebResourceResponse
     * @throws IOException if an error occurs (e.g. url doesn't point to any asset file).
     */
    @Throws(IOException::class)
    fun serve(context: Context, url: String): WebResourceResponse {
        // fetch assets path without leading slash
        val uri = Uri.parse(url)
        var assetPath = uri.path!!
        if (assetPath.startsWith("/")) {
            assetPath = assetPath.substring(1)
        }

        var loadFromAssets = false
        if (assetPath.startsWith(ASSETS_PREFIX)) {
            loadFromAssets = true
            assetPath = assetPath.substring(ASSETS_PREFIX.length)
        }

        // interpret url with trailing slashes as referring to directory with index.html
        if (assetPath.isEmpty() || assetPath.endsWith("/")) {
            assetPath += "index.html"
        }

        var mimeType:String? = null
        val assetStream: InputStream
        assetStream = if (assetPath == "favicon.ico") {
            LOGGER.debug("favicon.ico request found. Returning empty response.")
            ByteArrayInputStream(ByteArray(0))
        } else {
            LOGGER.debug("Fetching asset '{}' ...", assetPath)
            if (loadFromAssets) {
                context.assets.open(assetPath)
            }
            else {
                val assetResource = runBlocking {
                    FirebaseFacade.fetchAboutResource(assetPath)
                }
                mimeType = assetResource?.mimeType
                val assetBytes = assetResource?.let {
                    if (it.base64Encoded) {
                        Base64.decode(it.data, Base64.DEFAULT)
                    } else {
                        it.data.toByteArray(Charsets.UTF_8)
                    }
                }
                if (assetBytes != null) {
                    ByteArrayInputStream(assetBytes)
                } else {
                    // for failed html requests, send 404
                    if (assetPath.endsWith(".html")) {
                        context.assets.open("404/index.html")
                    } else {
                        // send empty responses for failed non html requests.
                        ByteArrayInputStream(ByteArray(0))
                    }
                }
            }
        }
        if (mimeType == null || mimeType == "") {
            mimeType = getMimeType(assetPath)
        }
        LOGGER.debug("Returning {} response from {}...", mimeType, assetPath)
        return WebResourceResponse(
            mimeType,
            "utf-8", assetStream
        )
    }

    fun resolveUrl(
        rootRelativeUrl: String,
        queryParams: Array<String?>?
    ): String {
        // ensure there is no leading slash.
        var rootRelativeUrl = rootRelativeUrl
        if (rootRelativeUrl.startsWith("/")) {
            rootRelativeUrl = rootRelativeUrl.substring(1)
        }

        // ensure launch url has trailing slash
        val launchUrl = StringBuilder(LAUNCH_URL)
        if (launchUrl[launchUrl.length - 1] != '/') {
            launchUrl.append("/")
        }
        launchUrl.append(rootRelativeUrl)
        if (queryParams != null) {
            var i = 0
            while (i < queryParams.size) {
                val q = queryParams[i]
                var v: String? = null
                if (i + 1 < queryParams.size) {
                    v = queryParams[i + 1]
                }
                if (i > 0) {
                    launchUrl.append("&")
                } else {
                    launchUrl.append("?")
                }
                launchUrl.append(Uri.encode(q))
                if (v != null) {
                    launchUrl.append("=")
                    launchUrl.append(Uri.encode(v))
                }
                i += 2
            }
        }
        return launchUrl.toString()
    }

    internal fun getMimeType(path: String): String {
        val fileExtension = getFileExtension(path).toLowerCase()
        return when (fileExtension) {
            "html" -> "text/html"
            "css" -> "text/css"
            "js" -> "text/javascript"
            "png" -> "image/png"
            "gif" -> "image/gif"
            "jpg", "jpeg" -> "image/jpeg"
            "svg" -> "image/svg+xml"
            "woff" -> "application/font-woff"
            "eot" -> "application/vnd.ms-fontobject"
            "ttf", "otf" -> "application/font-sfnt"
            "ico" -> "image/vnd.microsoft.icon"
            else -> MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension) ?: "*/*"
        }
    }

    internal fun getFileExtension(path: String): String {
        val lastPeriodIndex = path.lastIndexOf('.')
        return if (lastPeriodIndex != -1) {
            path.substring(lastPeriodIndex + 1)
        } else ""
    }
}